[flang] AliasAnalysis: Handle fir.load on fir.alloca

jdenny-ornl opened the pull request:
For example, determine that the address in p below cannot alias the address of v:

```
subroutine test()
  real, pointer :: p
  real, target :: t
  real :: v
  p => t
  v = p
end subroutine test
```

jdenny-ornl committed:
[flang] AliasAnalysis: Handle fir.load on fir.alloca

For example, determine that the address in p below cannot alias the
address of v:

```
subroutine test()
  real, pointer :: p
  real, target :: t
  real :: v
  p => t
  v = p
end subroutine test
```

Renaud-K commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -497,6 +482,29 @@ static Value getPrivateArg(omp::BlockArgOpenMPOpInterface &argIface,
   return privateArg;
 }
 
+/// Temporary function to skip through all the no op operations
```
Can you not move and change the function at the same time? It is harder to see what the actual change is. If it is a function that is defined before, you can just provide a declaration instead.

jdenny-ornl replied:
Done.

jdenny-ornl committed:
Move getOriginalDef back to where it was, as requested

jdenny-ornl committed:
Apply clang-format

Renaud-K commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -522,6 +546,12 @@ AliasAnalysis::Source AliasAnalysis::getSource(mlir::Value v,
+          // (alias-analysis-2.fir), but does flang currently generate such
+          // code?
+          if (isPointerReference(ty))
+            attributes.set(Attribute::Pointer);
```
Can you try remove the 2 changes you made for alias-analysis-2.fir, here and above and set the pointer attribute in
the case for fir::BoxAddrOp:

```
        .Case<fir::BoxAddrOp>([&](auto op) {
          v = op->getOperand(0);
          defOp = v.getDefiningOp();
          if (fir::isPointerType(v.getType()))
            attributes.set(Attribute::Pointer);
          if (mlir::isa<fir::BaseBoxType>(v.getType()))
            followBoxData = true;
        })
```

jdenny-ornl replied:
Thanks, check-flang passes with that change.  But can you help me to understand the following points?

1.  In alias-analysis-2.fir, `p1.addr` is `fir.alloca !fir.ptr<f32>`.   I understood that to be pointer non-data (i.e., address of a pointer not of its target).  After applying your suggestion, getSource labels it as non-pointer data instead.  How is the latter correct?
2. As this PR's todo comment asks, when does flang generate something like `fir.alloca !fir.ptr<f32>`?  What semantics does it represent in Fortran?
3. This PR's approach of adding `attributes.set(Attribute::Pointer)` to getSource's AllocaOp case seems consistent with the existing AddrOfOp and dummy arg cases (the two cases that getSource already handles after LoadOp).  Your suggestion makes AllocaOp inconsistent with AddrOfOp and dummy args.  Why is that inconsistency correct?

jdenny-ornl replied:
Thanks for your reply.  In this thread, we've discussed three changes to AliasAnalysis.cpp:

- C1: [unset followingData on AllocaOp](https://github.com/jdenny-ornl/llvm-project/blob/00aeefe3a5fb1944b7b6be0098edb1916d77be9a/flang/lib/Optimizer/Analysis/AliasAnalysis.cpp#L546-L556) (included in the PR)
- C2: [add Attribute::Pointer to AllocaOp](https://github.com/jdenny-ornl/llvm-project/blob/00aeefe3a5fb1944b7b6be0098edb1916d77be9a/flang/lib/Optimizer/Analysis/AliasAnalysis.cpp#L570-L575) (included in the PR)
- C3: [add Attribute::Pointer to BoxAddrOp](https://github.com/llvm/llvm-project/pull/117785#discussion_r1859666490) (your suggestion)

There are two checks in alias-analysis-2.fir that sometimes break depending on which of the above changes we have while keeping all remaining changes from this PR:

- T1: [p1.addr#0 <-> p2.addr#0: NoAlias](https://github.com/jdenny-ornl/llvm-project/blob/00aeefe3a5fb1944b7b6be0098edb1916d77be9a/flang/test/Analysis/AliasAnalysis/alias-analysis-2.fir#L9)
- T2: [boxp1.addr#0 <-> func.region0#0: MayAlias](https://github.com/jdenny-ornl/llvm-project/blob/00aeefe3a5fb1944b7b6be0098edb1916d77be9a/flang/test/Analysis/AliasAnalysis/alias-analysis-2.fir#L34C1-L34C56)

T2 breaks when none of the three changes are applied because boxp1.addr is not treated as a pointer.  Each of C2 and C3 fixes that.

T1 breaks when C2 is applied without C1 (regardless of whether C3 is applied) because C2 causes the address of a pointer without a box to be mislabeled as pointer data instead of pointer non-data.

So, for check-flang to pass, we seem to need either C1+C2 (PR as it is now), C3 alone, or C1+C2+C3.  I am concerned about C3 alone due to points 1 and 3 in [my previous comment](https://github.com/llvm/llvm-project/pull/117785#discussion_r1860967421).  I don't know what cases C1+C2+C3 handles that C1+C2 doesn't, but we should extend the test suite to cover them if we go with that.

> 1. We introduced data vs non-data, to lift ambiguity when following a box and to distinguish between the box's address or the address it wraps.

Understood.

> There is no such ambiguity for scalars, so the classical way of distinguishing between 2 sources should work.

I'm not sure which case you are referring to when you mention scalars. As far as I know, the relevant cases (T1 and T2) involve addresses of/in pointers, sometimes with boxes.

> 2. After discussing with Jean, all the pointers are boxed since HLFIR, but to optimize the code some may be reverted back to just allocas. It is best to keep options open.

OK, so it seems we have to keep at least one of the combinations listed above.

> 3. You may keep the change at the lines 553,554 (where this discussion is anchored), it is not harmful and consistent with the global case, as you point out.

That's C2.  I think you're suggesting we go with C2+C3, but that breaks T1, as described above.

Renaud-K replied:
I had no issues with C2+C3. I no longer have a workspace with your changes. Can you double check ?

jdenny-ornl replied:
I pushed the C2+C3 solution to the PR (temporarily).  I've marked C1, C2, C3, T1, and T2 with comments.  We should see T1 fail in CI.

Renaud-K replied:
C2 is the issue. Unlike `!fir_ptr<T>`, a `!fir_ref <!fir_ptr<T>>` is not a fortran `POINTER`. I am aware of the discrepancy with fir.AddrOf and as you can see there is a TODO there. I will be revisiting the whole `fir.load` approach to address it.

jdenny-ornl replied:
> C2 is the issue. Unlike `!fir_ptr<T>`, a `!fir_ref <!fir_ptr<T>>` is not a fortran `POINTER`.

I have been assuming that `!fir.ref<!fir.ptr<T>>` is the address of a pointer (that is, pointer non-data) but just missing the box.  Do you agree?

> I am aware of the discrepancy with fir.AddrOf

And dummy args.

> and as you can see there is a TODO there. I will be revisiting the whole `fir.load` approach to address it.

What do you have planned, and what's your timeline?

Renaud-K replied:
> I have been assuming that !fir.ref<!fir.ptr<T>> is the address of a pointer (that is, pointer non-data) but just missing the box. Do you agree?
I don't want to introduce data vs non-data for scalars if we do not have to. I do not think in those terms. I try to rely on the original concepts and see if they can be leveraged. They can be leveraged. We have 2 different allocas, they are not `POINTER`s or `TARGET`s, they cannot alias. Period. There is not need to introduce new concepts to solve this problem.

My timeline: ideally shortly after you commit this patch, while the iron is still hot.

jdenny-ornl replied:
Are you using the term scalar to describe `!fir.ref<!fir.ptr>`?

Renaud-K replied:
The box is an aggregate (a non-scalar). 
The definition of non-data is: a memory reference to a box.  We do not need to change this definition.

jdenny-ornl replied:
If you're opposed to C1 because it expands the definition of non-data, the only remaining solution we've discussed is C3 alone.  I've pushed a commit to the PR to leave C3 uncommented but to comment out C1 and C2.

If we go with C3 alone, what happens when an address is loaded from within `fir.alloca !fir.ptr<T>`?  Without C2, how will it pick up a pointer attribute so we can detect aliasing properly?  C3 doesn't help because there's no box and so no BoxAddrOp.

I've extended alias-analysis-2.fir with a new T3 to show such a case.  It fails with C3 alone but passes when using C1+C2.  I don't know if this case is realistic because I don't know how to make flang generate `fir.alloca !fir.ptr<T>`.

jdenny-ornl replied:
I've restored C1+C2 so that all tests pass again.

Given that you're opposed to C1 but we so far have no other solution that doesn't [cause apparent regressions](https://github.com/llvm/llvm-project/pull/117785#discussion_r1924405040), what do you feel is the path forward?

Renaud-K replied:
T3 and T4 of course belong in a separate PR along with all discussions around changing the definition of data versus non-data. I can see that there would be much to discuss about changing this definition and we would need time to think things through. You have been provided with a solution and very reasonable and actionable feedback to move forward.

jdenny-ornl replied:
After talking to @jeanPerier, you said `fir.alloca !fir.ptr<T>` needs to be supported, so we should avoid the potential behavior regression for it that T1 checks.  Also for `fir.alloca !fir.ptr<T>`, this PR with C3 alone will cause other behavior regressions, which T3 and T4 check.

It seems you have decided it will be fine to introduce only the latter behavior regressions when landing this PR and to fix them in a later PR.  To that end, I have adjusted this PR to have your C3 alone and to add fixme comments for T3 and T4.

For the sake of future development, would you please comment here on why it is reasonable to permit only the latter behavior regressions for now?

Thanks for your help.

Renaud-K replied:
OK. I was given a few days to go deeper into this. I am not sure I agree with your conclusion on T3 but T4 is an issue.
T4 is a new test, so not a regression but would be if added before these changes. And that's what I am trying to say. These changes could really be broken up into separate PRs. It would help considerably with this review. 

Also, as I am looking into this more, would you like us to considering `fir.alloca !fir.ptr<T>` as non-data and `!fir.ptr<T>`  as data?

jdenny-ornl replied:
> OK. I was given a few days to go deeper into this.

Sounds good.  Thanks.

> I am not sure I agree with your conclusion on T3 but T4 is an issue.

We can return to T3 after discussing the path forward for T4.

> T4 is a new test, so not a regression but would be if added before these changes. And that's what I am trying to say.

Whether a behavior is tested upstream does not indicate whether breaking it is a regression.  @jeanPerier has specifically [expressed concern](https://github.com/llvm/llvm-project/pull/94242#pullrequestreview-2248272387) about lack of test coverage for Flang's alias analysis.

> These changes could really be broken up into separate PRs. It would help considerably with this review.

I'm fine to split it up, but I'm striving to land the PRs in an order that doesn't leave anything important broken for an indefinite amount of time.  If we agree T3 or T4 reveals a regression, then we need to figure out what PR to propose as a *parent* of this PR.

> Also, as I am looking into this more, would you like us to considering `fir.alloca !fir.ptr<T>` as non-data and `!fir.ptr<T>` as data?

Yes, if `fir.alloca !fir.ptr<T>` and thus `!fir.ref<!fir.ptr<T>>` are important cases and essentially represent the same Fortran concepts as `fir.alloca !fir.box<!fir.ptr<T>>` and `!fir.ref<!fir.box<!fir.ptr<T>>>`, then that seems reasonable to me.

I tried that approach in this PR originally with C1, but my inability to make Flang produce those cases from Fortran means my implementation is not well tested, and I'm not confident I've handled all cases.

Renaud-K replied:
So T3 is the only test that would require this change. We are going to take it separately. It will be running along with T4 after this PR: https://github.com/llvm/llvm-project/pull/126110. I have explained it to @jeanPerier and @vzakhari, and  taken it apart will make for a much simpler discussion. 

I have 2 additional PRs that I would like to commit. One is https://github.com/llvm/llvm-project/pull/125925 and another one to come. They should be vastly no ops and mostly clean-ups that will help with your change. Mostly, I would like to take advantage that my head is fully into this code base now to take advantage of the time spent. They are changes that I have been meaning to do for awhile with @jeanPerier and @vzakhari as well and will address the TODO in `fir::AddrOfOp`

jdenny-ornl replied:
> So T3 is the only test that would require this change.

I think you mean that C1+C2 relative to main will not affect T1, T2, and T4 but will affect T3 (MayAlias->NoAlias).  We still have to agree on whether that's an improvement.  Of course, this PR introduces many other tests that focus on this PR's original goal, and those are affected.

> We are going to take it separately. It will be running along with T4 after this PR: #126110. I have explained it to @jeanPerier and @vzakhari, and taken it apart will make for a much simpler discussion.

Sure, that's fine.  Will that discussion happen in one of the other PRs you are planning?

Renaud-K replied:
After my changes, C1 to C3 will probably look very different. It will not make sense to discuss them.
Could be best to start a new PR.

As far as discussing T3. You should be driving the discussion maybe even file an issue. But because of the time spent, my management is now asking for benchmark data so be ready for that.

The 80+ other original tests look good. We can move forward with them. But this should just be a matter of generalizing what we are doing in the box load since, from what I could tell, it is what is limiting these tests. We do need to change our definition of data vs non-data for them.

jdenny-ornl replied:
> After my changes, C1 to C3 will probably look very different. It will not make sense to discuss them.  Could be best to start a new PR.
>
> As far as discussing T3. You should be driving the discussion maybe even file an issue. But because of the time spent, my management is now asking for benchmark data so be ready for that.

I have data to show how this PR plus one more I haven't proposed yet significantly improved performance of an app we are investigating. The data is stale by so many months now, I should probably run it again before posting it.  I haven't proposed the other PR yet because I suspect I'll have to rewrite it by the time this PR lands, as your comment above also suggests.

However, the vast majority of the discussion for this PR (including C1, C3, T1, T2, T3, and T4) has been for `fir.alloca !fir.ptr<T>` and `!fir.ref<!fir.ptr<T>>`, a case that, so far in this discussion, no one has made Flang produce for any Fortran program.  But you and @jeanPerier say it matters.  If we need to justify all the time we're spending on this PR, perhaps one of you can show when that case matters.  I certainly don't have benchmark data for it.

> The 80+ other original tests look good. We can move forward with them. But this should just be a matter of generalizing what we are doing in the box load since, from what I could tell, it is what is limiting these tests. We do need to change our definition of data vs non-data for them.

Sounds good.  Thanks for your help.

jdenny-ornl replied:
Thanks to PR #126156, the main change in this PR no longer appears to affect `fir.alloca !fir.ptr<T>` and `!fir.ref<!fir.ptr<T>>`.  Thus, C1+C2 or C3 is no longer needed to avoid breaking T1 and T2, which are no longer interesting and don't need to be marked in the test suite. Finally, @Renaud-K copied this PR's T3 and T4 to his PR #125917.

I'm resolving this and related conversations so we can focus on the rest of this PR.

jdenny-ornl committed:
Merge branch 'main' into flang-aa-load-alloca

jdenny-ornl committed:
Merge branch 'main' into flang-aa-load-alloca

jdenny-ornl commented:
ping

Renaud-K commented:
I posted comments a while back, as you can see from the date, but they remained in "Pending" status. My bad.

jdenny-ornl committed:
Show the effect of C2+C3

As discussed at
<https://github.com/llvm/llvm-project/pull/117785#discussion_r1905907812>,
it breaks a test, marked as T1.

jdenny-ornl committed:
Show an issue with C3 alone

Add a failing test marked as T3.

jdenny-ornl committed:
Further demonstrate issue with C3 alone

Add another failing test marked as T4.

Renaud-K commented on flang/test/Analysis/AliasAnalysis/alias-analysis-2.fir:
```diff
@@ -47,6 +50,14 @@
 // CHECK-DAG: arg2.load#0 <-> arg2.addr#0: MustAlias
 // CHECK-DAG: boxp1.addr#0 <-> arg2.addr#0: MayAlias
 
+// The address in a pointer cannot alias the address of a pointer, even when the
```
It would have to be addressed in a separate MR.

jdenny-ornl replied:
C3 alone introduces regressions for these cases:

For T3:

- Without this PR: MayAlias
- With C1+C2: NoAlias (right)
- With C3 alone: MustAlias (wrong)

For T4:
- Without this PR: MayAlias
- With C1+C2: MayAlias
- With C3 alone: NoAlias (wrong)

If these case are important, we should fix them before landing this PR.

jdenny-ornl replied:
Resolving [as discussed in the other conversation](https://github.com/llvm/llvm-project/pull/117785#discussion_r1953308159).

jdenny-ornl committed:
Merge branch 'main' into flang-aa-load-alloca

jdenny-ornl committed:
Restore C1+C2 so all tests pass again

jdenny-ornl committed:
Merge branch 'main' into flang-aa-load-alloca

jdenny-ornl committed:
Remove C1+C2, restore C3, add test fixmes for regressions

Renaud-K commented on flang/test/Analysis/AliasAnalysis/load-ptr-alloca.fir:
Please rename this file with an `alias-analysis-` prefix so we can run all these tests with `lit --filter alias-analysis `

jdenny-ornl replied:
There are multiple existing tests in that directory that don't start with the prefix `alias-analysis-`.  The prefix is redundant because the directory's name is AliasAnalysis.  Use `lit --filter AliasAnalysis`.

jdenny-ornl committed:
Revert "Remove C1+C2, restore C3, add test fixmes for regressions"

This reverts commit 1ce92161e1bc1032fa9e28165686ccdf639f214c.

jdenny-ornl committed:
Merge branch 'main' into flang-aa-load-alloca

jdenny-ornl commented:
For now, I'm just trying to keep this PR up to date with main so it's easy to reference.  I understand it might need to be rewritten after @Renaud-K lands his PRs.

jdenny-ornl committed:
Merge branch 'main' into flang-aa-load-alloca

jdenny-ornl committed:
Clean up alias-analysis-2.fir comments

Don't introduce T1 and T2 comments as I'm not sure they matter so much
going forward.  As for T3, make it clear what T4 references.

kawashima-fj commented:
@jdenny-ornl After this commit, [a test which uses Cray pointer](https://github.com/fujitsu/compiler-test-suite/blob/main/Fortran/0734/0734_0018.f90) fails. Could you take a look?

Before this commit (660cdace559a):

```console
$ flang Fortran/0734/0734_0018.f90
$ ./a.out
 pass
```

After this commit (eb8ffd617a06):

```console
$ flang Fortran/0734/0734_0018.f90
$ ./a.out
 erorr-1
 pass
```

jdenny-ornl commented:
@kawashima-fj Thanks for the report, and sorry for the trouble.  I am able to reproduce the issue on a local test machine, and I will investigate.  If I don't find a simple resolution today, I will try to revert the change.  Let me know if you feel I should revert sooner.  Thanks.

jdenny-ornl commented:
> @jdenny-ornl After this commit, [a test which uses Cray pointer](https://github.com/fujitsu/compiler-test-suite/blob/main/Fortran/0734/0734_0018.f90) fails. Could you take a look?

I don't have experience working with cray pointers or forall loops in fortran.  However, what I've read suggests the above test is invalid, and so the result is undefined.

Specifically, the test has:

```
pointer (j,x)
j=loc(a(2))
forall (i=1:3)
 x(i)=a(i)+1
end forall
```

Because `x(i)` is an alias for `a(i+1)`, the forall body performs `a(i+1)=a(i)+1`.  From what I've read, in order to enable optimizations, forall doesn't permit this kind of data dependency among iterations.

Furthermore, on my test system, both gfortran 11.4.0 and 13.2.0 produce the same result for this test as flang does with this PR merged.

Let me know if you believe flang's previous behavior for this test is the only correct behavior.

kawashima-fj commented:
@jdenny-ornl Thanks for your investigation. I will check with my colleague who is familiar with the specifications.

kawashima-fj commented:
@jdenny-ornl I and my colleague confirmed. As you saied, it's an invalid test. I'm sorry to bother you.

jdenny-ornl commented:
@kawashima-fj No worries.  Thanks for confirming.
